src/napi-macros-extensions.h

```diff
@@ -41,3 +41,47 @@
   NAPI_STATUS_THROWS(napi_create_string_utf8(env, name, NAPI_AUTO_LENGTH, &return_value)); \
   free(name); \
   return return_value;
+
+#define NAPI_ASYNC_CARRIER_BEGIN(name) \
+  typedef struct name##_carrier_t { \
+    napi_ref callback_ref; \
+    napi_async_work async_work; \
+    dcn_context_t* dcn_context; \
+
+#define NAPI_ASYNC_CARRIER_END(name) \
+  } name##_carrier_t; \
+
+#define NAPI_ASYNC_EXECUTE(name) \
+  static void name##_execute(napi_env env, void* data)
+
+#define NAPI_ASYNC_GET_CARRIER(name) \
+  name##_carrier_t* carrier = (name##_carrier_t*)data; \
+
+#define NAPI_ASYNC_COMPLETE(name) \
+  static void name##_complete(napi_env env, napi_status status, void* data) \
+
+#define NAPI_ASYNC_CALL_AND_DELETE_CB() \
+  napi_value global; \
+  NAPI_STATUS_THROWS(napi_get_global(env, &global)); \
+  napi_value callback; \
+  NAPI_STATUS_THROWS(napi_get_reference_value(env, carrier->callback_ref, &callback)); \
+  NAPI_STATUS_THROWS(napi_call_function(env, global, callback, argc, argv, NULL)); \
+  NAPI_STATUS_THROWS(napi_delete_reference(env, carrier->callback_ref)); \
+  NAPI_STATUS_THROWS(napi_delete_async_work(env, carrier->async_work)); \
+
+#define NAPI_ASYNC_NEW_CARRIER(name) \
+  name##_carrier_t* carrier = calloc(1, sizeof(name##_carrier_t)); \
+  carrier->dcn_context = dcn_context; \
+
+#define NAPI_ASYNC_QUEUE_ASYNC_WORK_AND_RETURN(name, cb) \
```

Comment:
Maybe call it `NAPI_ASYNC_QUEUE_WORK()` and make `NAPI_RETURN_UNDEFINED()` explicit? Functions might need to return something, even though they do async work.